.github/workflows/tsan.yaml

```diff
@@ -0,0 +1,112 @@
+            libffi-dev liblzma-dev
+      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
+        with:
+          path: jax
+      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
+        with:
+          repository: python/cpython
+          path: cpython
+          ref: v3.13.1
+      - name: Build CPython with TSAN enabled
+        run: |
+          cd cpython
+          mkdir ${GITHUB_WORKSPACE}/cpython-tsan
+          CC=clang-18 CXX=clang++-18 ./configure --prefix ${GITHUB_WORKSPACE}/cpython-tsan --disable-gil --with-thread-sanitizer
+          make -j64
+          make install
+          # Check whether free-threading mode is enabled
+          PYTHON_GIL=0 ${GITHUB_WORKSPACE}/cpython-tsan/bin/python3 -c "import sys; assert not sys._is_gil_enabled()"
+          ${GITHUB_WORKSPACE}/cpython-tsan/bin/python3 -m venv ${GITHUB_WORKSPACE}/venv
+
+          # Create archive to be used with bazel as hermetic python:
+          cd ${GITHUB_WORKSPACE} && tar -czpf python-tsan.tgz cpython-tsan
+      - name: Build and install JAX
+        run: |
+          source ${GITHUB_WORKSPACE}/venv/bin/activate
+          cd jax
+
+          export PYTHON_SHA256=($(sha256sum ${GITHUB_WORKSPACE}/python-tsan.tgz))
+          echo "Python sha256: ${PYTHON_SHA256}"
+
+          python build/build.py build --wheels=jaxlib \
+            --bazel_options=--repo_env=HERMETIC_PYTHON_VERSION=3.13-ft \
+            --bazel_options=--repo_env=HERMETIC_PYTHON_URL="file://${GITHUB_WORKSPACE}/python-tsan.tgz" \
+            --bazel_options=--repo_env=HERMETIC_PYTHON_SHA256=${PYTHON_SHA256} \
+            --bazel_options=--repo_env=HERMETIC_PYTHON_PREFIX="cpython-tsan/" \
+            --bazel_options=--color=yes \
+            --bazel_options=--copt=-fsanitize=thread \
+            --bazel_options=--linkopt="-fsanitize=thread" \
+            --bazel_options=--@rules_python//python/config_settings:py_freethreaded="yes" \
+            --clang_path=/usr/bin/clang-18
+      - name: Run tests
+        timeout-minutes: 120
+        env:
+          JAX_NUM_GENERATED_CASES: 1
+          JAX_ENABLE_X64: true
+          JAX_SKIP_SLOW_TESTS: true
+          PY_COLORS: 1
+        run: |
+          cd jax
+          echo "JAX_NUM_GENERATED_CASES=$JAX_NUM_GENERATED_CASES"
+          echo "JAX_ENABLE_X64=$JAX_ENABLE_X64"
+          echo "JAX_SKIP_SLOW_TESTS=$JAX_SKIP_SLOW_TESTS"
+
+          # As we do not have yet free-threading support
+          # there will be the following warning:
+          # RuntimeWarning: The global interpreter lock (GIL) has been enabled to load module 'jaxlib.utils',
+          # which has not declared that it can run safely without the GIL.
+          # To avoid that we temporarily define PYTHON_GIL
+          export PYTHON_GIL=0
+
+          # Set symlink to the bazel executable
+          bazel_exec=($(ls bazel-*))
+          ln -s ${bazel_exec} bazel
+
+          ./bazel test \
```

Comment:
We can then build and run the tests with `bazel` like this:

```
bazel test //tests:cpu_tests --test_env=TSAN_OPTIONS=halt_on_error=1,suppressions=/path/to/suppressions --test_env=JAX_TEST_NUM_THREADS=8
``` 

perhaps with a couple of the other environment variables.

Reply:
I would split build and test steps

Reply:
No, you shouldn't, because that's actually slower.